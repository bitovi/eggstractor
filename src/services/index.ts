export * from './collection.service';
export * from './token.service';
export * from './variable.service'; 