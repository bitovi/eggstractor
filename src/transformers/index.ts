export * from "./scss.transformer";
export * from "./css.transformer";
