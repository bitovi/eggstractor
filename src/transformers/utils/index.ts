export * from './deduplicate-messages.utils';
export * from './get-style-property-and-value.utils';
export * from './group-by.utils';
export * from './update-padding-based-on-border.utils';
