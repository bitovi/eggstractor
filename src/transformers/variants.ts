export type Input = {
  variants: Record<string, string>;
  css: Record<string, string>;
}[];

let i = 0;
/**
 * Unique numeric id's starting from 0
 */
const getId = () => {
  return ++i;
};

/**
 * Check if objects match all property and value pairs
 */
function shallowEqual<T extends Record<string, any>>(a: T, b: T): boolean {
  const keysA = Object.keys(a);
  const keysB = Object.keys(b);

  if (keysA.length !== keysB.length) {
    return false;
  }

  for (const key of keysA) {
    if (a[key] !== b[key]) {
      return false;
    }
  }

  return true;
}

type StyleNode = {
  cssProperty: string;
  cssValue: string;
  variants: Record<string, string>;
  possibleVariants: Record<string, string>;
  id: number;
};

/**
 * Filter any styles with unique variant combinations.
 * There is an exception, if a variant combinations is more common than all others, it will be considered "unique"
 */
const splitByMatch = (items: StyleNode[]): [StyleNode[], StyleNode[]] => {
  const matchingCounts = new Array<number>(items.length).fill(0);
  const exceptionCounts = new Array<number>(items.length).fill(0);

  for (let i = 0; i < items.length; i++) {
    for (let j = i + 1; j < items.length; j++) {
      const a = items[i];
      const b = items[j];
      if (
        a.cssProperty === b.cssProperty &&
        shallowEqual(a.variants, b.variants)
      ) {
        if (a.cssValue === b.cssValue) {
          exceptionCounts[i] += 1;
          exceptionCounts[j] += 1;
        } else {
          matchingCounts[i] += 1;
          matchingCounts[j] += 1;
        }
      }
    }
  }

  const matching: StyleNode[] = [];
  const nonMatching: StyleNode[] = [];

  for (let i = 0; i < items.length; i++) {
    // TODO: Allows for "shorter" classNames, but leads to "random" classNames
    // that are short but should be included in another className
    // Alternative is commented out below
    if (matchingCounts[i] && matchingCounts[i] > exceptionCounts[i]) {
      // TODO: Avoids "random" classNames that are short that should be included in existing classNames
      // but creates "longer" classNames
    // if (matchingCounts[i]) {// && matchingCounts[i] > exceptionCounts[i]) {
      matching.push(items[i]);
    } else {
      nonMatching.push(items[i]);
    }
  }

  return [nonMatching, matching];
};
/**
 * Compare element with every other element in array. If comparison is true, exclude from array
 */
// TODO: is this just Array.filter()
function removeByComparison<T>(
  items: T[],
  comparison: (a: T, b: T) => boolean,
): T[] {
  const result: T[] = [];

  for (let i = 0; i < items.length; i++) {
    const current = items[i];
    const isAlreadyIncluded = result.some((existing) =>
      comparison(existing, current),
    );

    if (!isAlreadyIncluded) {
      result.push(current);
    }
  }

  return result;
}

// TODO: consideration, we already have "Style" token, so maybe the shape of the input should match the existing style tokens
export const getInitialStyleNodes = (source: Input): StyleNode[] => {
  const styleNodes: StyleNode[] = [];

  for (const instance of source) {
    Object.entries(instance.css).map(([cssProperty, cssValue]) => {
      // Each node requires referring its "origin" node
      // The "origin" nodes are created in this loop for a single style's variants
      const id = getId();

      for (let i = 0; i < Object.keys(instance.variants).length; i++) {
        styleNodes.push({
          cssProperty,
          cssValue,
          variants: {},
          possibleVariants: { ...instance.variants },
          id,
        });
      }
    });
  }

  return styleNodes;
};

const getFinalizedStyleNodes = (
  styles: StyleNode[],
): [StyleNode[], StyleNode[]] => {
  // "uniques" are finalized, duplicates need to be processed further
  const [uniques, duplicates] = splitByMatch(styles);

  // Remove any variants trying to apply a style that has already been covered
  const cleanedDuplicates: StyleNode[] = duplicates.filter(
    (duplicate) => !uniques.some((unique) => unique.id === duplicate.id),
  );

  return [uniques, cleanedDuplicates];
};

export const createChildStyleNodes = (styles: StyleNode[]): StyleNode[] => {
  if (!styles.length) {
    return [];
  }

  const [uniques, duplicates] = getFinalizedStyleNodes(styles);

  const nestedChildStyleNodes = duplicates
    .flatMap((instance) => {
      return Object.entries(instance.possibleVariants).map(
        ([variantProperty, variantValue]) => {
          const possibleVariants = { ...instance.possibleVariants };
          delete possibleVariants[variantProperty];

          const styleNode: StyleNode = {
            cssProperty: instance.cssProperty,
            cssValue: instance.cssValue,
            variants: {
              ...instance.variants,
              [variantProperty]: variantValue,
            },
            possibleVariants,
            id: instance.id,
          };

          return styleNode;
        },
      );
    });

  const cleanedNestedChildStyleNodes: StyleNode[] = removeByComparison(
    nestedChildStyleNodes,
    (a, b) => {
      return (
        // Has to also match css style
        a.cssProperty === b.cssProperty &&
        a.cssValue === b.cssValue &&
        // Check if combination of variants match
        // areVariantsEqual(a, b) &&
        shallowEqual(a.variants, b.variants) &&
        // Check if style belongs to the same "instance"
        // TODO: we could do this by adding an id to instances instead
        shallowEqual(a.possibleVariants, b.possibleVariants)
      );
    },
  );

  return [...uniques, ...createChildStyleNodes(cleanedNestedChildStyleNodes)];
};

export const convertStyleNodesToCssStylesheet = (
  styleNodes: StyleNode[],
): Record<string, Record<string, string>> => {
  const styleTags: Record<string, Record<string, string>> = {};

  // Group all styles by "className/mixin/utility"
  for (const style of styleNodes) {
    const key = Object.entries(style.variants)
      .map(
        ([variantProperty, variantValue]) =>
          `${variantProperty}-${variantValue}`,
      )
      .join("--") || 'ROOT';

    styleTags[key] ??= {};
    if (
      styleTags[key][style.cssProperty] &&
      styleTags[key][style.cssProperty] !== style.cssValue
    ) {
      console.error(
        styleTags,
        key,
        styleTags[key],
        styleTags[key][style.cssProperty],
        style.cssValue,
      );
      throw new Error("Unexpected style exists for combination of variant");
    }
    styleTags[key][style.cssProperty] = style.cssValue;
  }

  return styleTags;
};

export const generateStyles = (source: Input) => {
  const initialStyleNodes = getInitialStyleNodes(source);
  const styleNodes = createChildStyleNodes(initialStyleNodes);
  const stylesheet = convertStyleNodesToCssStylesheet(styleNodes);
  return stylesheet;
};
