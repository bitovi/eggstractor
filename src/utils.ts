export default {
  toBase64(str: string): string {
    const base64chars = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/';
    const utf8str = encodeURIComponent(str).replace(/%([0-9A-F]{2})/g,
      (_, p1) => String.fromCharCode(parseInt(p1, 16))
    );
    let i = 0;
    let result = '';
    while (i < utf8str.length) {
      const char1 = utf8str.charCodeAt(i++);
      const char2 = i < utf8str.length ? utf8str.charCodeAt(i++) : NaN;
      const char3 = i < utf8str.length ? utf8str.charCodeAt(i++) : NaN;
      const enc1 = char1 >> 2;
      const enc2 = ((char1 & 3) << 4) | (char2 >> 4);
      const enc3 = ((char2 & 15) << 2) | (char3 >> 6);
      const enc4 = char3 & 63;
      result += base64chars[enc1] + base64chars[enc2] +
        (isNaN(char2) ? '=' : base64chars[enc3]) +
        (isNaN(char3) ? '=' : base64chars[enc4]);
    }
    return result;
  },
  sanitizeName(name: string): string {
    return name.toLowerCase().replace(/[^a-z0-9]/g, '-');
  },
  sanitizeSegment(segment: string): string {
    return segment.toLowerCase().replace(/[^a-z0-9]/g, '-');
  },
  rgbToHex(r: number, g: number, b: number): string {
    const rHex = Math.round(r * 255).toString(16).padStart(2, '0');
    const gHex = Math.round(g * 255).toString(16).padStart(2, '0');
    const bHex = Math.round(b * 255).toString(16).padStart(2, '0');
    return '#' + rHex + gHex + bHex;
  },
  rgbaToString(r: number, g: number, b: number, a: number): string {
    const rHex = Math.round(r * 255).toString(16).padStart(2, '0');
    const gHex = Math.round(g * 255).toString(16).padStart(2, '0');
    const bHex = Math.round(b * 255).toString(16).padStart(2, '0');
    const aHex = Math.round(a * 255).toString(16).padStart(2, '0');
    
    return a === 1 ? 
      '#' + rHex + gHex + bHex : 
      '#' + rHex + gHex + bHex + aHex;
  },
  groupBy<T>(arr: T[], key: (item: T) => string): Record<string, T[]> {
    return arr.reduce((groups, item) => {
      const k = key(item);
      if (!groups[k]) groups[k] = [];
      groups[k].push(item);
      return groups;
    }, {} as Record<string, T[]>);
  }
};