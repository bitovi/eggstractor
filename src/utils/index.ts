export * from './color.utils';
export * from './create-naming-convention'
export * from './constants.utils';
export * from './delay.utils';
export * from './get-valid-stylesheet-format.utils';
export * from './node.utils';
export * from './string.utils';
export * from './test.utils';
export * from './units.utils';
export * from './value.utils';
