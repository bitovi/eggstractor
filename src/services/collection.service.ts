import { ComponentSetToken, ComponentToken, TokenCollection } from '../types';
import { getProcessorsForNode } from '../processors';
import { extractComponentSetToken, extractComponentToken, extractNodeToken } from '../services';
import { getNodePathNames } from '../utils/node.utils';

function getFlattenedValidNodes(node: BaseNode): BaseNode[] {
  const result: BaseNode[] = [];

  function traverse(currentNode: BaseNode) {
    // Skip VECTOR and INSTANCE nodes entirely. INSTANCE nodes which are
    // instances of components, which are not relevant for token extraction.
    if ('type' in currentNode && ['VECTOR', 'INSTANCE'].includes(currentNode.type)) {
      return;
    }

    result.push(currentNode);

    if ('children' in currentNode) {
      for (const child of currentNode.children) {
        traverse(child);
      }
    }
  }

  traverse(node);
  return result;
}

export async function collectTokens(onProgress: (progress: number, message: string) => void) {
  const collection: TokenCollection = { tokens: [], components: {}, componentSets: {} };

  let componentToken: ComponentToken | null = null;
  let componentSetToken: ComponentSetToken | null = null;

  let totalNodes = 0;
  let processedNodes = 0;

  let lastTimestamp = Date.now();
  let lastPercentage = -1;

  async function processNode(node: BaseNode) {
    processedNodes++;

    // calculate integer percent in your 10–95 range
    const currentPercentage = Math.floor((processedNodes / totalNodes) * 85) + 10;
    const shouldUpdate = currentPercentage !== lastPercentage || processedNodes === totalNodes;

    if (shouldUpdate) {
      lastTimestamp = currentPercentage;
      onProgress(currentPercentage, `Processing nodes… ${processedNodes}/${totalNodes}`);

      // throttle yields to at most once every 200 ms
      const now = Date.now();
      if (now - lastTimestamp >= 200) {
        await new Promise((r) => setTimeout(r, 0));
        lastTimestamp = now;
      }
    }

    if ('type' in node && 'boundVariables' in node) {
      if (node.type === 'COMPONENT_SET') {
        componentSetToken = extractComponentSetToken(node);
        collection.componentSets[node.id] = componentSetToken;
      }

      if (node.type === 'COMPONENT') {
        componentToken = extractComponentToken(node, componentSetToken!);
        collection.components[node.id] = componentToken;
      }

      const nodePathNames = getNodePathNames(node);
      const processors = getProcessorsForNode(node);

      for (const processor of processors) {
        const tokens = await extractNodeToken(
          node,
          processor,
          nodePathNames,
          componentToken,
          componentSetToken,
        );
        collection.tokens.push(...tokens);
      }
    }
  }

  onProgress(0, 'Loading pages...');
  await figma.loadAllPagesAsync();

  onProgress(5, 'Counting nodes...');
  const validNodes = figma.root.children.flatMap((page) => getFlattenedValidNodes(page));
  totalNodes = validNodes.length;

  onProgress(10, `Processing ${totalNodes} nodes...`);

  // Process all pages in parallel for maximum speed
  const pagePromises = figma.root.children.map(async (page) => {
    const validNodes = getFlattenedValidNodes(page);

    for (const node of validNodes) {
      await processNode(node);
    }
  });

  await Promise.all(pagePromises);

  return collection as Readonly<TokenCollection>;
}
