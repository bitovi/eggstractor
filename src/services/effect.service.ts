import { VariableToken, TokenCollection } from '../types';
import { sanitizeName } from '../utils';

/**
 * Convert a Figma EffectStyle to a VariableToken
 */
function createEffectStyleToken(effectStyle: EffectStyle): VariableToken | null {
  try {
    // Convert effects array to CSS box-shadow value
    const boxShadowValues = effectStyle.effects
      .filter((effect) => effect.type === 'DROP_SHADOW' || effect.type === 'INNER_SHADOW')
      .map((effect) => {
        const { offset, radius, color, spread } = effect;
        const isInset = effect.type === 'INNER_SHADOW';

        // Convert RGBA to CSS rgba()
        const colorValue = `rgba(${Math.round(color.r * 255)}, ${Math.round(color.g * 255)}, ${Math.round(color.b * 255)}, ${color.a ?? 1})`;

        // Build box-shadow value: [inset] offset-x offset-y blur-radius [spread-radius] color
        const parts = [];
        if (isInset) parts.push('inset');
        parts.push(`${offset?.x || 0}px`);
        parts.push(`${offset?.y || 0}px`);
        parts.push(`${radius || 0}px`);
        if (spread) parts.push(`${spread}px`);
        parts.push(colorValue);

        return parts.join(' ');
      });

    if (boxShadowValues.length === 0) {
      return null;
    }

    const rawValue = boxShadowValues.join(', ');

    return {
      type: 'variable',
      path: [{ name: effectStyle.name, type: 'FRAME' }],
      property: 'box-shadow',
      name: sanitizeName(effectStyle.name),
      value: `$${sanitizeName(effectStyle.name)}`,
      rawValue,
      valueType: null,
      metadata: {
        figmaId: effectStyle.id,
        variableName: effectStyle.name,
        variableTokenType: 'primitive',
      },
    };
  } catch (error) {
    console.warn(`Failed to create token for effect style ${effectStyle.name}:`, error);
    return null;
  }
}

/**
 * Collect all Figma Effect Styles (box-shadow, etc.)
 */
export async function collectAllFigmaEffectStyles(
  collection: TokenCollection,
  onProgress: (progress: number, message: string) => void,
) {
  try {
    onProgress(7, 'Collecting Figma Effect Styles...');
    const effectTokens: VariableToken[] = [];

    // Get all local effect styles
    const effectStyles = await figma.getLocalEffectStylesAsync();

    for (const effectStyle of effectStyles) {
      onProgress(8, `Processing effect style: ${effectStyle.name}`);

      const token = createEffectStyleToken(effectStyle);
      if (token) {
        effectTokens.push(token);
      }
    }

    console.log(`âœ¨ Collected ${effectTokens.length} effect style tokens`);
    collection.tokens.push(...effectTokens);
  } catch (error) {
    console.warn('Failed to collect Figma Effect Styles:', error);
  }
}
