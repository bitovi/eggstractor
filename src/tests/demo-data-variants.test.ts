import { collectTokens } from '../services';
import { transformToScss } from '../transformers';
import { createTestData } from '../utils/test.utils';
import testDataDemo from './fixtures/figma-test-data_demo.json';
import * as variants from '../transformers/variants';

describe('Demo Data (real world-ish example for variant combinations)', () => {
  it('should process all demo data correctly with variant combinations', async () => {
    // Force feature flag for variant combinations to true for new behavior
    jest.spyOn(variants, 'USE_VARIANT_COMBINATION_PARSING').mockReturnValueOnce(true);
    
    const { setupTest } = createTestData(testDataDemo);
    const testSetup = await setupTest();

    global.figma = testSetup.figma;

    const tokens = await collectTokens(jest.fn());
    const { result } = transformToScss(tokens);

    expect(result).toMatchSnapshot('demo-data-variants');
  });
});
