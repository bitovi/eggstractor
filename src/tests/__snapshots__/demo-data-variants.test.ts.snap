// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Demo Data (real world-ish example for variant combinations) should process all demo data correctly with variant combinations: demo-data-variants 1`] = `
"// Generated SCSS Variables
$base-size-xs: 0.125rem;
$base-colour-grey-400: #282828;
$base-font-size-xl: 1.25rem;
$global-font-uidefaultfont: inter;
$base-font-linespacing-lg: 1.5rem;
$base-font-weight-semibold: 31.25rem;
$base-colour-static-white: #ffffff;
$base-size-lg: 0.75rem;
$base-radius-sm: 0.25rem;
$global-size-border-width: 0.062rem;
$base-colour-grey-100: #d9d9d9;
$global-colour-textonlight: #282828;
$base-font-weight-regular: 25rem;
$base-colour-grey-300: #747474;
$base-colour-blue-300: #0177cc;

// Generated SCSS Mixins
@mixin input {
  display: flex;
  flex-direction: column;
  align-items: flex-start;
  gap: $base-size-xs;
  padding: 0rem;
  width: 15.937rem;
}
@mixin input-label {
  display: flex;
  flex-direction: column;
  color: $base-colour-grey-400;
  font-family: $global-font-uidefaultfont;
  font-size: $base-font-size-xl;
  font-weight: 500;
  font-style: normal;
  line-height: $base-font-linespacing-lg;
  width: 15.937rem;
}
@mixin input-label-active {
  color: $base-colour-blue-300;
}
@mixin input-input-box {
  display: flex;
  flex-direction: row;
  align-items: center;
  gap: $base-size-lg;
  padding: 0.687rem $base-size-lg;
  background: $base-colour-static-white;
  width: 15.937rem;
  border-radius: $base-radius-sm;
}
@mixin input-input-box-resting {
  border: $global-size-border-width solid $base-colour-grey-100;
  box-shadow: inset 0 $global-size-border-width 0 0 $base-colour-grey-100, inset (-$global-size-border-width) 0 0 0 $base-colour-grey-100, inset 0 (-$global-size-border-width) 0 0 $base-colour-grey-100, inset $global-size-border-width 0 0 0 $base-colour-grey-100;
}
@mixin input-input-box-hover {
  border: $global-size-border-width solid $base-colour-grey-300;
  box-shadow: inset 0 $global-size-border-width 0 0 $base-colour-grey-300, inset (-$global-size-border-width) 0 0 0 $base-colour-grey-300, inset 0 (-$global-size-border-width) 0 0 $base-colour-grey-300, inset $global-size-border-width 0 0 0 $base-colour-grey-300;
}
@mixin input-input-box-active {
  border: 0.125rem solid $base-colour-blue-300;
  box-shadow: inset 0 0.125rem 0 0 $base-colour-blue-300, inset -0.125rem 0 0 0 $base-colour-blue-300, inset 0 -0.125rem 0 0 $base-colour-blue-300, inset 0.125rem 0 0 0 $base-colour-blue-300;
}
@mixin input-input-box-input-text {
  display: flex;
  flex-direction: column;
  color: $global-colour-textonlight;
  font-family: $global-font-uidefaultfont;
  font-size: 1rem;
  font-weight: 400;
  font-style: normal;
  line-height: 20px;
  width: 14.437rem;
}
"
`;
